balancer/pickfirst_leaf/pickfirst_leaf.go

```diff
@@ -0,0 +1,640 @@
+			ConnectivityState: connectivity.Connecting,
+			Picker:            &picker{err: balancer.ErrNoSubConnAvailable},
+		})
+		b.requestConnection()
+	} else if b.state == connectivity.Idle {
+		// If this is not the first resolver update and we're in IDLE, remain
+		// in IDLE until the picker is used.
+		b.cc.UpdateState(balancer.State{
+			ConnectivityState: connectivity.Idle,
+			Picker:            &idlePicker{exitIdle: b.ExitIdle},
+		})
+	} else if b.state == connectivity.TransientFailure {
+		// If we're in TRANSIENT_FAILURE, we stay in TRANSIENT_FAILURE until
+		// we're READY. See A62.
+		b.requestConnection()
+	}
+	return nil
+}
+
+// UpdateSubConnState is unused as a StateListener is always registered when
+// creating SubConns.
+func (b *pickfirstBalancer) UpdateSubConnState(subConn balancer.SubConn, state balancer.SubConnState) {
+	b.logger.Errorf("UpdateSubConnState(%v, %+v) called unexpectedly", subConn, state)
+}
+
+func (b *pickfirstBalancer) Close() {
+	b.serializer.TrySchedule(func(_ context.Context) {
+		b.close()
+	})
+	<-b.serializer.Done()
+}
+
+// close closes the balancer.
+// Only executed in the context of a serializer callback.
+func (b *pickfirstBalancer) close() {
+	b.serializerCancel()
+	b.closeSubConns()
+	b.state = connectivity.Shutdown
+}
+
+// ExitIdle moves the balancer out of idle state. It can be called concurrently
+// by the idlePicker and clientConn so access to variables should be synchronized.
+func (b *pickfirstBalancer) ExitIdle() {
+	b.serializer.TrySchedule(func(_ context.Context) {
+		b.exitIdle()
+	})
+}
+
+// exitIdle starts a connection attempt if not already started.
+// Only executed in the context of a serializer callback.
+func (b *pickfirstBalancer) exitIdle() {
+	if b.state == connectivity.Idle {
+		b.requestConnection()
+	}
+}
+
+func (b *pickfirstBalancer) closeSubConns() {
+	for _, sd := range b.subConns.Values() {
+		sd.(*scData).subConn.Shutdown()
+	}
+	for _, k := range b.subConns.Keys() {
+		b.subConns.Delete(k)
+	}
+}
+
+// deDupAddresses ensures that each address belongs to only one endpoint.
+func deDupAddresses(endpoints []resolver.Endpoint) []resolver.Endpoint {
+	seenAddrs := resolver.NewAddressMap()
+	newEndpoints := []resolver.Endpoint{}
+
+	for _, ep := range endpoints {
+		addrs := []resolver.Address{}
+		for _, addr := range ep.Addresses {
+			if _, ok := seenAddrs.Get(addr); ok {
+				continue
+			}
+			addrs = append(addrs, addr)
+		}
+		if len(addrs) == 0 {
+			continue
+		}
+		newEndpoints = append(newEndpoints, resolver.Endpoint{
+			Addresses:  addrs,
+			Attributes: ep.Attributes,
+		})
+	}
+	return newEndpoints
+}
+
+// shutdownRemaining shuts down remaining subConns. Called when a subConn
+// becomes ready, which means that all other subConn must be shutdown.
+func (b *pickfirstBalancer) shutdownRemaining(selected *scData) {
+	for _, v := range b.subConns.Values() {
+		sd := v.(*scData)
+		if sd.subConn != selected.subConn {
+			sd.subConn.Shutdown()
+		}
+	}
```

Comment:
We are doing this pattern in a couple of places where we traverse the addressMap and do *some* processing on each entry. Does it make sense to add a method on the addressMap type which would traverse the map and invoke a given function for each entry? Does not have to part of this PR. We can file a TODO/issue if you think its a reasonable idea.

Reply:
As long as we are not in the RPC path (i.e picker, transport etc), we shouldn't worry excessively about performance. I do agree that it doesn't save too many lines of code for us here. But I was simply wondering if having such functionality in the `AddressMap` type would be helpful in general.

Good to leave this as is. Thanks.